include/zephyr/llext/elf.h

```diff
@@ -0,0 +1,445 @@
+	elf32_word r_info;
+};
+
+/**
+ * @brief Relocation symbol index from r_info
+ *
+ * @param i Value of r_info
+ */
+#define ELF32_R_SYM(i) ((i) >> 8)
+
+/**
+ * @brief Relocation type from r_info
+ *
+ * @param i Value of r_info
+ */
+#define ELF32_R_TYPE(i) ((i) & 0xff)
+
+/**
+ * @brief Relocation entry for 64-bit ELFs
+ */
+struct elf64_rel {
+	/** Offset in section to perform a relocation */
+	elf64_addr r_offset;
+	/** Information about relocation, related symbol and type */
+	elf64_xword r_info;
+};
+
+/** @brief Relocation symbol from r_info
+ *
+ * @param i Value of r_info
+ */
+#define ELF64_R_SYM(i) ((i) >> 32)
+
+/**
+ * @brief Relocation type from r_info
+ *
+ * @param i Value of r_info
+ */
+#define ELF64_R_TYPE(i) ((i) & 0xffffffff)
+
+#define R_386_NONE 0
+#define R_386_32 1
+#define R_386_PC32 2
+#define R_386_GOT32 3
+#define R_386_PLT32 4
+#define R_386_COPY 5
+#define R_386_GLOB_DAT 6
+#define R_386_JMP_SLOT 7
+#define R_386_RELATIVE 8
+#define R_386_GOTOFF 9
+
+#define R_ARM_NONE 0
+#define R_ARM_PC24 1
+#define R_ARM_ABS32 2
+#define R_ARM_REL32 3
+#define R_ARM_COPY 4
+#define R_ARM_CALL 28
+#define R_ARM_V4BX 40
+
+#define R_XTENSA_NONE 0
+#define R_XTENSA_32 1
+#define R_XTENSA_SLOT0_OP 20
+
+/**
+ * @brief Program header(32-bit)
+ */
+struct elf32_phdr {
+	elf32_word p_type;
+	elf32_off p_offset;
+	elf32_addr p_vaddr;
+	elf32_addr p_paddr;
+	elf32_word p_filesz;
+	elf32_word p_memsz;
+	elf32_word p_flags;
+	elf32_word p_align;
+};
+
+/**
+ * @brief Program header(64-bit)
+ */
+struct elf64_phdr {
+	elf64_word p_type;
+	elf64_off p_offset;
+	elf64_addr p_vaddr;
+	elf64_addr p_paddr;
+	elf64_xword p_filesz;
+	elf64_xword p_memsz;
+	elf64_word p_flags;
+	elf64_xword p_align;
+};
+
+/**
+ * @brief Program segment type
+ */
+#define PT_LOAD 1
+
+/**
+ * @brief Dynamic section entry(32-bit)
+ */
+struct elf32_dyn {
+	elf32_sword d_tag;
+	union {
+		elf32_word d_val;
+		elf32_addr d_ptr;
+	} d_un;
+};
+
+/**
+ * @brief Dynamic section entry(64-bit)
+ */
+struct elf64_dyn {
+	elf64_sxword d_tag;
+	union {
+		elf64_xword d_val;
+		elf64_addr d_ptr;
+	} d_un;
+};
+
+#ifdef CONFIG_64BIT
+typedef struct elf64_ehdr elf_ehdr_t;
+typedef struct elf64_shdr elf_shdr_t;
+typedef struct elf64_phdr elf_phdr_t;
+typedef elf64_addr elf_addr;
+typedef elf64_half elf_half;
+typedef elf64_xword elf_word;
+typedef struct elf64_rela elf_rel_t;
+typedef struct elf64_sym elf_sym_t;
+#define ELF_R_SYM ELF64_R_SYM
+#define ELF_R_TYPE ELF64_R_TYPE
+#define ELF_ST_BIND ELF64_ST_BIND
+#define ELF_ST_TYPE ELF64_ST_TYPE
+#else
+typedef struct elf32_ehdr elf_ehdr_t;
+typedef struct elf32_shdr elf_shdr_t;
+typedef struct elf32_phdr elf_phdr_t;
+typedef elf32_addr elf_addr;
+typedef elf32_half elf_half;
+typedef elf32_word elf_word;
+typedef struct elf32_rel elf_rel_t;
+typedef struct elf32_sym elf_sym_t;
+#define ELF_R_SYM ELF32_R_SYM
+#define ELF_R_TYPE ELF32_R_TYPE
+#define ELF_ST_BIND ELF32_ST_BIND
+#define ELF_ST_TYPE ELF32_ST_TYPE
+#endif
```

Comment:
not sure what you want to do with these w.r.t doxygen comments? Also note that the typedefs in the `CONFIG_64BIT` section won't show up until you do `#if defined(CONFIG_64BIT) || defined(__DOXYGEN__)`

Reply:
Nah that's fine, please don't bother :)

Reply:
I'm annotating all the typedefs with *very* simple doc comments